fjord/feedback/models.py

```diff
@@ -645,100 +632,103 @@ class PostResponseSerializer(serializers.Serializer):
-    campaign = NoNullsCharField(max_length=100, required=False, default=u'')
+    source = serializers.CharField(
+        max_length=100, allow_null=False, default=u'')
+    campaign = serializers.CharField(
+        max_length=100, allow_null=False, default=u'')
 
-    def validate_url(self, attrs, source):
-        value = attrs[source]
+    def validate_url(self, value):
         if value:
             if not is_url(value):
-                msg = u'{0} is not a valid url'.format(value)
-                raise serializers.ValidationError(msg)
-
-        return attrs
+                raise serializers.ValidationError(
+                    u'{0} is not a valid url'.format(value)
+                )
+        return value
 
-    def validate_product(self, attrs, source):
+    def validate_product(self, value):
         """Validates the product against Product model"""
-        value = attrs[source]
-
         # This looks goofy, but it makes it more likely we have a
         # cache hit.
         products = Product.objects.values_list('display_name', flat=True)
         if value not in products:
             raise serializers.ValidationError(
-                '{0} is not a valid product'.format(value))
-        return attrs
-
-    def restore_object(self, attrs, instance=None):
-        # Note: instance should never be anything except None here
-        # since we only accept POST and not PUT/PATCH.
-
-        opinion = Response(
-            happy=attrs['happy'],
-            url=attrs['url'].strip(),
-            description=attrs['description'].strip(),
-            category=attrs['category'].strip(),
-            product=attrs['product'].strip(),
-            channel=attrs['channel'].strip(),
-            version=attrs['version'].strip(),
-            platform=attrs['platform'].strip(),
-            locale=attrs['locale'].strip(),
-            manufacturer=attrs['manufacturer'].strip(),
-            device=attrs['device'].strip(),
-            country=attrs['country'].strip(),
-            source=attrs['source'].strip(),
-            campaign=attrs['campaign'].strip(),
-            api=1,  # Hard-coded api version number
-        )
-
-        # If there's a user agent, infer all the things from the user
-        # agent.
-        user_agent = attrs['user_agent'].strip()
+                u'{0} is not a valid product'.format(value)
+            )
+        return value
+
+    def create(self, validated_data):
+        # First, pop out 'email' since it's not in the Response.
+        validated_data.pop('email', None)
+
+        # Strip string data
+        for key in ['url', 'description', 'category', 'product',
+                    'channel', 'version', 'platform', 'locale',
+                    'manufacturer', 'device', 'country', 'source',
+                    'campaign', 'user_agent']:
+            if key in validated_data:
+                validated_data[key] = validated_data[key].strip()
```

Comment:
While working on APIs in SUMO for BuddyUp, I ran into a lot of cases where behavior was encoded only in the views, which produced a lot of annoying validation problems. Because of that, I am going to try and push things into the models in SUMO.

This seems appropriate for the models. Maybe in a save method? Or maybe these fields on the model should be a subclass of CharField that automatically strips?